active/0006-extensibility-support-for-deployment-stacks.md

```diff
@@ -0,0 +1,821 @@
+
+Bicep may provide function expressions to use in place of object literal secure property references that will compile to
+the literal format in the ARM template. One such example is the existing `kv.getSecret(...)` function.
+
+For the "Resource" source, custom properties ("credentialType" in the example) on this object will be determined by the
+extension. In a future REP, there will be a design for how extensions can provide this detail. For the time being it
+will be hard coded in the engine to accept certain properties based on the resource type of the resource supplied. This
+future REP will off load the responsibility of how to get the secret from the user and Deployment engine to the
+extension implementer.
+
+**For stack deployments, any language expressions present in extension configurations must be constrained to prevent
+persisting user secrets long-term and causing security risks. There are no restrictions for non-stack deployments.**
+Secure properties will be re-fetched on demand for stack deployments. Non-secure property expressions will be resolved
+at deployment time and the value of the expression will be persisted long-term for stack deployments.
+
+For stack deployments, Bicep expressions prohibited anywhere inside extension configurations:
+- Secure deployment parameters
+- `keyVault.getSecret(...)`  (this expresses a key vault reference rather than a value)
+
+For stack deployments, additional prohibited expressions for non-secure properties:
+- Value access to inherited extension config secure properties
+
+The reasoning behind the disallowed expressions for stack deployments is because the extension configuration properties
+will be evaluated by the deployment engine and the evaluation results will be returned in the deployment GET response.
+The stacks service will persist this data long-term and use it for the next stack update or deletion that requires
+extension resource deletions. **The deployments backend will need to ensure future sources of certain secure data are 
+blocked from being used for stack deployment extension configurations.**
+
+The stacks service already sends a stack resource ID HTTP request header to the deployments service. This will be used
+to enable additional validation in the Deployment engine to enforce these expression rules. It will also enable the
+return of extension configurations in the Deployment GET response for API versions that support it. This header is only
+needed for the initial deployment PUT, as its value will be saved on the deployment metadata and this will be used to
+drive the API behavior. For deployments that are not initiated by the stack service, the configuration portion of the
+extension details will not be returned, as no additional validation on language expressions is done, which means we
+cannot guarantee compliance to the ARM RP RPC rule that a resource GET cannot return secret data.
+
+#### Stacks mode for Bicep files
```

Comment:
Should we consider alternative mechanisms for declaring that the file represents a stack? Should we introduce something in the .bicep files to let users indicate it?

Reply:
I ended up deciding to stick to linter rules for this validation, thus the designation is in the bicepconfig.

My very first idea was to include a directive line in the file, something like `compatibility stacks`. These `compatibility` directives would only apply limitations, so it would be compatible with a deployment flow without them. It didn't seem necessary to introduce this given bicepconfig can scope the linting rules to a set of files and within the files, the linter rule can be manually ignored. Having linting rule(s) that default to informational will enable the user to see the considerations but also ignore them or enforce them.